### PR TITLE
Implemented Binary Search Tree as Data Structure. Implemented with Iterator Interface.

### PR DESCRIPTION
**Pull request by Ramy-Badr-Ahmed**
**Description**

The BST allows for efficient data organization, enabling operations like insertion, deletion, and traversal in logarithmic time complexity, making it ideal for scenarios where fast lookups, insertions, and deletions are necessary. It can be utilized in various applications, including:

- Database indexing: Quickly locating and managing records.
- Memory management: Allocating and deallocating memory efficiently.
- Autocompletion: Facilitating rapid searches for suggestions based on user input.
- Sorting algorithms: Implementing efficient sorting methods.

**Contents**

- `BSTNode` Class: Represents individual nodes in the Binary Search Tree with attributes `key`, `value`, `left`, `right` and `parent`. Where:
        `key`: The value used to organize the nodes in the tree.
        `value`: Stores any associated data for the node.
        `left`, `right`: Pointers to the left and right child nodes.
        `parent`: Pointer to the parent node.

    The `BSTNode` class encapsulates the structure of each node, linking to its child and parent nodes to maintain the tree structure and facilitate traversal.

- `BSTree` Class: Implements the core functionalities of a Binary Search Tree.

     `insert()`: Adds a new node to the tree and return the tree root. If duplicated, throws a `DuplicateKeyException`.
     `search()`: Locates and returns a node with a specific key or returns null if not exists.
     `remove()`: Deletes and returns a node from the tree, updates the tree referencing, restructures the tree to ensure it remains valid. If node not exists, returns null.
      `minNode()`: Return the minimum node in the BST, useful in node removal (in-order successor swap).
     `getHeight()`: Get the height of the given node relative to the farthest leaf node.
     `getDepth()`: Get the depth of the given node relative to the root of the tree.
     `serialize()`: Converts the segment tree into a JSON string.
     `deserialize()`: Restores it from the serialized format.

- `BinaryTreeTraversal` Class:  
    - Provides traversal methods:

       `inOrder()`: In-order traversal (left-root-right).
       `preOrder()`: Pre-order traversal (root-left-right).
       `postOrder()`: Post-order traversal (left-right-root).
       `breadthFirst()`: Breadth-first (level-order) traversal.

    - Implements the `Iterator` interface Methods:

        `current()`: Returns the current node in the iteration.
        `key()`: Returns the key of the current node.
        `next()`: Moves the pointer to the next node.
        `rewind()`: Resets the pointer to the first node.
        `valid()`: Checks if the current position is valid in the iteration.

     The Iterator allows looping through the Binary Search Tree by iterating over the nodes in `inOrder`, `preOrder`, or `postOrder` manner. Example:

```php
$bTree = new BSTree($data);       
foreach ($bTree as $node) {  
  // looping nodes in inOrder manner (default)
  // access $node-key and $node->value  
} 
 
//Alternatively, set another traversal method before looping by:
$bTree->setTraversalType('postOrder');
$bTree->setTraversalType('preOrder');

//or initalize tree with:
$bTree = new BSTree($data, 'postOrder');     
$bTree = new BSTree($data, 'preOrder');     
``` 

**Unit Tests**:

   `BSTreeTest.php`: Contains PHPUnit tests to validate the correct behavior of insertion, deletion, searching, traversing and iterating the BST, ensuring the integrity of the tree structure. It also consider edge cases for these operations as well as large trees.
   
**GitHub Actions**

All tests and workflows (in my forked repository) have passed.

- PHPCS

[Code style](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/11314666671/job/31464916144)

- DIRECTORY.md
 
[directory_md](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/11314666668/job/31464916131)

- build
 
[PHP Composer](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/11314666669/job/31464916152)
________

**Reference**

**[Data Structures and Algorithms in C++, 2nd Edition](https://www.wiley.com/en-br/Data+Structures+and+Algorithms+in+C%2B%2B%2C+2nd+Edition-p-9780470383278)**

### CODE REVIEW

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
Hi @darwinz 
Looking forward to your review 🙂

File: tests/DataStructures/BSTreeTest.php [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+        $root = $this->tree->getRoot();
+
+        $node150 = $this->tree->search(150);
+        $node110 = $this->tree->search(110);
+        $node70 = $this->tree->search(70);
+
+        $this->assertEquals(0, $this->tree->getdepth($root), "The root node should have a depth of 0.");
+        $this->assertEquals(1, $this->tree->getdepth($node150), "Node 150 should have a depth of 1.");
+        $this->assertEquals(4, $this->tree->getdepth($node110), "Node 110 should have a depth of 4.");
+        $this->assertEquals(6, $this->tree->getdepth($node70), "Node 300 should have a depth of 6.");
```
Comment (darwinz) [inline tests/DataStructures/BSTreeTest.php:323]:
```suggestion
        $this->assertEquals(6, $this->tree->getdepth($node70), "Node 70 should have a depth of 6.");
```

Comment (Ramy-Badr-Ahmed) [reply tests/DataStructures/BSTreeTest.php]: Thanks!
Just added and corrected more comments as well

**Review by darwinz — APPROVED** [review]
